README.md
# <b>Extensible Keys</b>
## Extensible Key Format:
The Extensible Key format was designed as an alternative method for storing and password-protecting keys, largely made for personal research purposes.
## <br>File Syntax:
### All .ekf files follow this format:
[Base64-Encoded 128-Bit Salt]:[Base64-Encoded 128-Bit IV]<br>
BEGIN KEYS:<br>[Key Data]<br>END KEYS!<br>CHECKSUM: [Base64-encoded SHA256 checksum]</b><br><br>
<b>128-Bit Salt:</b><br>
This randomly generated salt is SHA256 hashed with your password to create an AES256 key which is used both for checksum validation, and encrypting/decrypting keys.<br><br>
<b>128-Bit IV:</b><br>
The randomly generated Initilization Vector to encrypt/decrypt keys in AES CBC Mode<br><br>
<b>Key Data:</b><br>
Keys are stored in a relational database with titles (i.e., "RSA PUB", "THE VAULT*") The titles are largely arbitrary with two notes: <br>1) No colons are allowed as this would break the formatting of the file<br>
2) Titles for encrypted keys must end with an asterisk, as this tells the parser to encrypt/decrypt the associated  key before saving/returning them.<br>
Each new line in the file between "START KEYS:" and "STOP KEYS!" is counted as seperate entry in the key database, the proper syntax for one of these entries is as follows:
<b>[Key Title]: [Base64-Encoded (possibly encrypted) Key][*]</b>
<br><br>
<b>SHA256 Checksum:</b>
This SHA256 hash is created by hashing all of the keys together, and the  encrypting the resulting bytes. This checksum is checked when reading an .ekf file, by generating the same hash, and instead of decrypting the original, the newly generated hash is encrypted instead, this simultaneously autheticates the password, and verifies the integrity of the stored keys<br><br><br>

## API:
This repository includes a python modue entitled "ExtensibleKeys.py" this should be considered the reference implementation. This module includes four functions:

<br><b>write_file(file_name: str, keys: dict, pw: str):</b><br>
Creates .ekf files given the desired file path and name, dictionary of keys, and and a password. The dictionary is in this format: {title(str): key(bytes)}. The password will be utf-8 encoded. Using the name of a prexisting file will raise an error to prevent accidental overwriting.<br>
<br><b>read_file(file_name: str, pw: str) -> dict:</b><br>
Reads .ekf files, given a file name and password, raises a DecryptionError if the password is incorrect or if the checksum can't be passed.<br>
<br><b> append_keys(file_name: str, new_keys: dict, pw: str):</b><br>
 Appends keys to a file, performs exactly like write_file except for that it appends to an existing file as opposed to creating one.<br>
 <br> <b>delete_keys(file_name: str, target_keys: tuple, pw: str):</b><br>
 Will remove specified keys from a file. Raises a ValueError if a non-existant key is passed.

# Security considerations and other notes:
Implement standard password requirements for protecting these files. <br><br>
As with any other password, randomized passwords should be  implemented if possible<br><br>
The same error being used for any time decryption fails is intentional, as it prevents Padding Orace attacks. All future implementations should provide simmilar ambiguity in such cases.<br><br>
It's advisable to name your keys in such a way that makes their function non-obvious, ideally slowing down any attacker that may manage to gain access to the password for a large file<br><br>
These files are suceptible to brute-force attacks, so it's still <b>strongly</b> recomended to not publicize files with sensitive keys, even if they are encrypted.<br><br>
One could make an implementation of this file format where accessing a file with no encrypted keys (i.e., a list of PGP public keys) doesn't require a password, however this could be considered a security risk as the password is also used to validate the authenticity of the stored keys.








CHANGELOG.md
# Changelog
### v0.1.0:
- Initial release

### v0.2.0:
- Added append_keys and delete_key

### v0.2.1:
- read_keys will now include the asterisk in the title of an encrypted password, as removing it made append_keys decrypt keys before saving them again.<br><br>
- Restructured directory

### v0.2.2:
- Failure to decrypt now results in an error in ALL scenarios 


extensiblekeys.py
import hashlib
import base64
import os

from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad

# raised when an .ekf file cannot be parsed
class EkfFileError(Exception):
    def __init__(self, message: str = "Invalid .ekf file"):
        super().__init__(message)


# raised when a key is incorrect or a checksum is failed
class DecryptionError(ValueError):
    def __init__(self, message: str = 'Failed to decrypt file'):
        super().__init__(message)


# creates a SHA-256 hash of all the keys in a file with salt
def create_checksum(keys: dict, salt: bytes) -> bytes():
    checksum = bytes()
    for i in keys.values():
        checksum += i
    return hashlib.sha256(checksum+salt).digest()


# creates a file given a dictionary of keys and a password of any length
def write_file(file_name: str, keys: dict, pw: str):
    # validate file name
    if not file_name.endswith('.ekf'):
        raise ValueError('Invalid File Extension')
    elif os.path.exists(file_name):
        raise FileExistsError()
    # generate cryptographic information
    salt = get_random_bytes(16)
    iv = get_random_bytes(16)
    # a string to be written to the file as the final step of the function
    output = f'{base64.b64encode(salt).decode("ascii")}:{base64.b64encode(iv).decode("ascii")}\nSTART KEYS:\n'
    titles = []
    # generate an AES cipher from the given password and previously generated cryptographic info
    aes_key = hashlib.sha256(pw.encode("utf-8")+salt).digest()
    cipher = AES.new(hashlib.sha256(pw.encode("utf-8")+salt).digest(), AES.MODE_CBC, iv)
    for title in keys:
        # validate key title
        key = keys[title]
        if ':' in title or title == "START KEYS:\n" or title == "STOP KEYS!":
            raise ValueError(f"Invalid Key Title: {title}")
        elif title in titles:
            raise ValueError(f"Repeat Key Title: {title}")
        # encrypt key if asked and save it to output
        if "*" in title:
            data = cipher.encrypt(pad(key, 16))
        else:
            data = key
        output += f'{title}: {base64.b64encode(data).decode("ascii")}\n'
        titles.append(title)
    # generate and store checksum
    output += 'STOP KEYS!\nCHECKSUM: '
    output += base64.b64encode(cipher.encrypt(pad(create_checksum(keys, salt), 16))).decode("ascii")
    # save output 
    with open(file_name, 'w') as f:
        f.write(output)
   

# loads keys from  a file and unencrypts them as needed.
def read_file(file_name: str, pw: str) -> dict:
    # validate .ekf file
    if not file_name.endswith('.ekf'):
        raise EkfFileError()
    with open(file_name) as k: 
        key_dat = k.readlines()
        if key_dat[1] != 'START KEYS:\n' or key_dat[-2] != 'STOP KEYS!\n' or not key_dat[-1].startswith('CHECKSUM:'):
            raise EkfFileError()
    keys = {}
    salt, iv = key_dat[0].split(':')
    salt = base64.b64decode(salt.encode('ascii'))
    iv = base64.b64decode(iv.replace('\n', '').encode('ascii'))
    cipher = AES.new(hashlib.sha256(pw.encode("utf-8")+salt).digest(), AES.MODE_CBC, iv)
    for i in key_dat[2:]:
        if 'STOP KEYS!' in i:
            break
        title, key = i.split(': ')
        key = key.replace('\n', '')
        if '*' in title:
            # decrypt keys if required
            try:
                keys[title] = unpad(cipher.decrypt(base64.b64decode(key.encode("ascii"))), 16)
            # padding errors are presented as value errors, so this is used to make padding errors indistinguishable from
            # the error raised by using an incorrect key (with correct padding), or an invlaid checksum, removing, or at 
            # least mitigating the risk of padding oracle attacks 
            except ValueError: 
                raise DecryptionError()
            
        else: 
            keys[title] = base64.b64decode(key.encode('ascii'))
    # generate and validate a checksum, if it suceeds the function returns the keys, otherwise it raises an error
    checksum = create_checksum(keys, salt)
    check = key_dat[-1].replace('CHECKSUM: ', '')
    if unpad(cipher.decrypt(base64.b64decode(check.encode("ascii"))), 16) != checksum:
        raise DecryptionError
    return keys


# adds key(s) to a file without disrupting any of the current ones
def append_key(file_name: str, new_keys: dict, pw: str):
    keys =  read_file(file_name, pw)
    keys = keys | new_keys
    os.remove(file_name)
    write_file(file_name, keys, pw)


# removes key(s) from a file
def delete_keys(file_name: str, target_keys: tuple, pw: str):
    keys = read_file(file_name, pw)
    for i in target_keys:
        try:
            del keys[i]
        except KeyError:
            raise ValueError(f'Key {i} not found!')
    os.remove(file_name)
    write_file(file_name, keys, pw)
    


extensiblekeys/ekfutils.py
import hashlib
import base64
import os

from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes
from Crypto.Util.Padding import pad, unpad

# raised when an .ekf file cannot be parsed
class EkfFileError(Exception):
    def __init__(self, message: str = "Invalid .ekf file"):
        super().__init__(message)


# raised when a key is incorrect or a checksum is failed
class DecryptionError(ValueError):
    def __init__(self, message: str = 'Failed to decrypt file'):
        super().__init__(message)


# creates a SHA-256 hash of all the keys in a file with salt
def create_checksum(keys: dict, salt: bytes) -> bytes():
    checksum = bytes()
    for i in keys.values():
        checksum += i
    return hashlib.sha256(checksum+salt).digest()


# creates a file given a dictionary of keys and a password of any length
def write_file(file_name: str, keys: dict, pw: str):
    # validate file name
    if not file_name.endswith('.ekf'):
        raise ValueError('Invalid File Extension')
    elif os.path.exists(file_name):
        raise FileExistsError()
    # generate cryptographic information
    salt = get_random_bytes(16)
    iv = get_random_bytes(16)
    # a string to be written to the file as the final step of the function
    output = f'{base64.b64encode(salt).decode("ascii")}:{base64.b64encode(iv).decode("ascii")}\nSTART KEYS:\n'
    titles = []
    # generate an AES cipher from the given password and previously generated cryptographic info
    aes_key = hashlib.sha256(pw.encode("utf-8")+salt).digest()
    cipher = AES.new(hashlib.sha256(pw.encode("utf-8")+salt).digest(), AES.MODE_CBC, iv)
    for title in keys:
        # validate key title
        key = keys[title]
        if ':' in title or title == "START KEYS:\n" or title == "STOP KEYS!":
            raise ValueError(f"Invalid Key Title: {title}")
        elif title in titles:
            raise ValueError(f"Repeat Key Title: {title}")
        # encrypt key if asked and save it to output
        if "*" in title:
            data = cipher.encrypt(pad(key, 16))
        else:
            data = key
        output += f'{title}: {base64.b64encode(data).decode("ascii")}\n'
        titles.append(title)
    # generate and store checksum
    output += 'STOP KEYS!\nCHECKSUM: '
    output += base64.b64encode(cipher.encrypt(pad(create_checksum(keys, salt), 16))).decode("ascii")
    # save output 
    with open(file_name, 'w') as f:
        f.write(output)
   

# loads keys from  a file and unencrypts them as needed.
def read_file(file_name: str, pw: str) -> dict:
    # validate .ekf file
    if not file_name.endswith('.ekf'):
        raise EkfFileError()
    with open(file_name) as k: 
        key_dat = k.readlines()
        if key_dat[1] != 'START KEYS:\n' or key_dat[-2] != 'STOP KEYS!\n' or not key_dat[-1].startswith('CHECKSUM:'):
            raise EkfFileError()
    keys = {}
    salt, iv = key_dat[0].split(':')
    salt = base64.b64decode(salt.encode('ascii'))
    iv = base64.b64decode(iv.replace('\n', '').encode('ascii'))
    cipher = AES.new(hashlib.sha256(pw.encode("utf-8")+salt).digest(), AES.MODE_CBC, iv)
    for i in key_dat[2:]:
        if 'STOP KEYS!' in i:
            break
        title, key = i.split(': ')
        key = key.replace('\n', '')
        if '*' in title:
            # decrypt keys if required
            try:
                title = title.replace('*', '')
                keys[title] = unpad(cipher.decrypt(base64.b64decode(key.encode("ascii"))), 16)
            # padding errors are presented as value errors, so this is used to make padding errors indistinguishable from
            # the error raised by using an incorrect key (with correct padding), or an invlaid checksum, removing, or at 
            # least mitigating the risk of padding oracle attacks 
            except ValueError: 
                raise DecryptionError()
            
        else: 
            keys[title] = base64.b64decode(key.encode('ascii'))
    # generate and validate a checksum, if it suceeds the function returns the keys, otherwise it raises an error
    checksum = create_checksum(keys, salt)
    check = key_dat[-1].replace('CHECKSUM: ', '')
    if unpad(cipher.decrypt(base64.b64decode(check.encode("ascii"))), 16) != checksum:
        return 'File could not be decrypted'
    return keys


# adds key(s) to a file without disrupting any of the current ones
def append_key(file_name: str, new_keys: dict, pw: str):
    keys =  read_file(file_name, pw)
    keys = keys | new_keys
    os.remove(file_name)
    write_file(file_name, keys, pw)


# removes key(s) from a file
def delete_keys(file_name: str, target_keys: tuple, pw: str):
    keys = read_file(file_name, pw)
    for i in target_keys:
        try:
            del keys[i]
        except KeyError:
            raise ValueError(f'Key {i} not found!')
    os.remove(file_name)
    write_file(file_name, keys, pw)
    
